data_model/src/predicate.rs

```diff
@@ -3,38 +3,66 @@
 #[cfg(not(feature = "std"))]
 use alloc::vec;
 
+use core::fmt::Display;
+
 use super::*;
 use crate::{IdBox, Name, Value};
 
 /// Predicate combinator enum.
 #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub enum PredicateBox {
+pub enum GenericPredicateBox<P> {
```

Comment:
```suggestion
pub enum GenericPredicateBox<P> {
```

Should we enforce a trait bound?

Reply:
We can't without polluting everything downstream with explicit lifetimes, I think, since `PredicateTrait` is generic over a reference to input type, which is `&Value` for filter predicates. If we change `PredicateTrait` to be generic over input type directly (`Value`), this problem pops up for permission validators, which accept a tuple of references. And we can't make input associated type instead, since some predicates are applicable over all types implementing certain trait.

Constructor though can and should've enforce this, which I added.
In addition to that we could make the enum private and wrap it into an opaque struct, which would deny every possible way of instantiating `GenericPredicateBox` with a type not implementing `PredicateTrait`. I find access to enum variants potentially useful, so I guess I would be weakly opposed to such a solution.

Reply:
OK. I'd revisit this later. Leave a comment to that effect.